metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/sync/BlockSynchronizer.scala

```diff
@@ -0,0 +1,210 @@
+      _    <- persist(quorumCertificate.blockHash, path)
+    } yield ()
+
+  /** Download a block and all of its ancestors into the in-memory block store.
+    *
+    * Returns the path from the greatest ancestor that had to be downloaded
+    * to the originally requested block, so that we can persist them in that order.
+    *
+    * The path is maintained separately from the in-memory store in case another
+    * ongoing download would re-insert something on a path already partially removed
+    * resulting in a forest that cannot be traversed fully.
+    */
+  private def download(
+      from: A#PKey,
+      blockHash: A#Hash,
+      path: List[A#Hash]
+  ): F[List[A#Hash]] = {
+    storeRunner
+      .runReadOnly {
+        blockStorage.contains(blockHash)
+      }
+      .flatMap {
+        case true =>
+          path.pure[F]
+
+        case false =>
+          inMemoryStore
+            .runReadOnly {
+              blockStorage.get(blockHash)
+            }
+            .flatMap {
+              case Some(block) =>
+                downloadParent(from, block, path)
+
+              case None =>
+                getAndValidateBlock(from, blockHash)
+                  .flatMap {
+                    case Some(block) =>
+                      inMemoryStore.runReadWrite {
+                        blockStorage.put(block)
+                      } >> downloadParent(from, block, path)
+
+                    case None =>
+                      Timer[F].sleep(retryTimeout) >>
+                        download(from, blockHash, path)
+                  }
+            }
+      }
+  }
+
+  private def downloadParent(
+      from: A#PKey,
+      block: A#Block,
+      path: List[A#Hash]
+  ): F[List[A#Hash]] = {
+    val blockHash       = Block[A].blockHash(block)
+    val parentBlockHash = Block[A].parentBlockHash(block)
+    download(from, parentBlockHash, blockHash :: path)
+  }
+
+  /** Try downloading the block from the source and perform basic content validation. */
+  private def getAndValidateBlock(
+      from: A#PKey,
+      blockHash: A#Hash
+  ): F[Option[A#Block]] =
+    getBlock(from, blockHash).map { maybeBlock =>
+      maybeBlock.filter { block =>
+        Block[A].blockHash(block) == blockHash &&
+        Block[A].isValid(block)
+      }
+    }
+
+  /** See how far we can go in memory from the original block hash we asked for,
+    * which indicates the blocks that no concurrent download has persisted yet,
+    * then persist the rest.
+    *
+    * Only doing oinepersist operation at a time to make sure there's no competition
+    * in the insertion order of the path elements among concurrent downloads.
+    */
+  private def persist(
+      blockHash: A#Hash,
+      path: List[A#Hash]
+  ): F[Unit] =
+    semaphore.withPermit {
+      inMemoryStore
+        .runReadOnly {
+          blockStorage.getPathFromRoot(blockHash).map(_.toSet)
+        }
+        .flatMap { unpersisted =>
+          persist(path, unpersisted)
```

Comment:
Since this is guarded by a semaphore I think it should hold that `unpersisted.isEmpty || path.toSet == unpersisted`. So how about:

```
// no .map(_.toSet)
.flatMap {
  case Nil => ().pure[F]
  case unpersisted => persistAndClear(unpersisted)
}
```

I would use a different name for the inner function. At least my brain is easily tricked towards seeing a recursion.

Reply:
Sure, happy to rename, I had to look hard what the difference was now as well. I wouldn't move the `Nil` here though, it's the first thing the other method does anyway, it's okay to call it with an empty list, since that's what it does anyway. Prefer to have check of the same condition. 

`unpersisted` and `path` are not necessarily the same. Here's a scenario:

```
                   B5 - B7
                  /
B0* - B1* - B2 - B4 - B6
             \ 
              B3
```
Say that B1 is our last persisted (marked by `*`) block, and we have 3 proposal coming in: B3, B7 and B6, all enqueued for download at the same time. This should result in 3 paths: `[B2, B3]; [B2, B4, B5, B7]; [B2, B4, B6]`. These blocks will all be added to the in memory block store as we walk backwards toward the persistent part, ending at B1. 

Then we call `persist`, which will remove the blocks from the in-memory store and move them to persistent storage in the order they appear in the `path`. We only do one at a time because of the semaphore, but the order can be arbitrary. Say we start with the last path.

After persisting `[B2, B4, B6]` the in-memory store will not have any of them any more. Take the 2nd path next. We start by determining which blocks are still unpersisted by traversing from B7 back on the parent path in the in-memory tree; this gives us an `unpersisted` path of `[B5, B7]`, which is not the same as the original we had to traverse back to B2.

I can't remember why I wasn't using just `unpersisted` as a path though, why use it as a filter while traversing `path`. Maybe I didn't have the semaphore at the time. :thinking:

Reply:
Oh yeah, I know what I meant by this:
> The path is maintained separately from the in-memory store in case another ongoing download would re-insert something on a path already partially removed resulting in a forest that cannot be traversed fully.

While `persist` is protected by a semaphore, the concurrent downloads are free to _insert_ into the in-memory tree. I was thinking about a scenario where download 1 is persisting and removing block while download 2 is re-inserting them during traversal. 

Take the previous scenario. Say we started downloading B6 and B7 at the same time, from different sources, Alice and Bob, and due to timing what happens is that we start downloading B2 from Alice and Bob at the same time. Say Alice's finishes first and we start persisting the path `[B2, B4, B6]`: we remove B2 from the in-memory tree, persist it, then B4, then B6. During this time, the download of B2 from Bob finishes and is re-inserted into the in-memory tree.

When Bob considers what he can persist, the in memory tree has B2, B5 and B7 in it, so it's not a tree any more, but a forest, B4 is missing. If Bob only checked how far he has can traverse from B7 up the tree, he'd only persist B5 and B7, and not remove B2 from the in-memory store, leaking memory. 

That's why Bob will always try to persist `[B2, B4, B5, B7]` but be advised that only B5 and B7 actually need insertion into the persistent storage, B2 must already be there because it's unreachable, so B2 is just removed from the in-memory store, and B4 does nothing because it's not in there.

Reply:
Nice explanation, thank you